### PR TITLE
Expose API to specify custom context manager for checkpoint

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96866
* #96862
* __->__ #96783
* #90105

Per [design](https://docs.google.com/document/d/1v-yqRqiWA6dIUOw5OpqFs2PqSQIbDEkwRPGk9FcYnxg/edit) we want (1) to allow the user to pass in a function that returns two context managers (2) a per-call API only for now, and (3) do not upstream selective checkpoint for the short term.

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -165,7 +165,17 @@ def backward(ctx, *args):
...
-def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
+def noop_context_fn():
+    return contextlib.nullcontext(), contextlib.nullcontext()
+
+
+def checkpoint(
+    function,
+    *args,
+    use_reentrant: bool = True,
+    context_fn: Callable[[], Tuple[ContextManager, ContextManager]] = noop_context_fn,
```
Comment (albanD) [inline torch/utils/checkpoint.py:176]: Can we make this kwarg only?

**Review by albanD — APPROVED** [review]
Sounds good to me.
@fmassa does this solves your problem?

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py:176]: Already is :P https://stackoverflow.com/questions/64423083/why-cant-both-args-and-keyword-only-arguments-be-mixed-with-args-and-kwargs

**Pull request closed without merging**